cpp/include/cudf/ast/detail/transform.cuh

```diff
@@ -0,0 +1,423 @@
+                      lhs,
+                      rhs,
+                      output,
+                      op);
+    } else {
+      release_assert(false && "Invalid operator arity.");
+    }
+  }
+};
+
+/**
+ * @brief Evaluate an expression applied to a row.
+ *
+ * This function performs an n-ary transform for one row on one thread.
+ *
+ * @param evaluator The row evaluator used for evaluation.
+ * @param data_references Array of data references.
+ * @param operators Array of operators to perform.
+ * @param operator_source_indices Array of source indices for the operators.
+ * @param num_operators Number of operators.
+ * @param row_index Row index of data column(s).
+ */
+__device__ void evaluate_row_expression(detail::row_evaluator const& evaluator,
+                                        const detail::device_data_reference* data_references,
+                                        const ast_operator* operators,
+                                        const cudf::size_type* operator_source_indices,
+                                        cudf::size_type num_operators,
+                                        cudf::size_type row_index)
+{
+  auto operator_source_index = cudf::size_type(0);
+  for (cudf::size_type operator_index(0); operator_index < num_operators; operator_index++) {
+    // Execute operator
+    auto const op = operators[operator_index];
+    evaluate_row_operator_functor{}(
+      op, evaluator, data_references, operator_source_indices, operator_source_index, row_index);
+  }
+}
+
+/**
+ * @brief Kernel for evaluating an expression on a table to produce a new column.
+ *
+ * This evaluates an expression over a table to produce a new column. Also called an n-ary
+ * transform.
+ *
+ * @tparam block_size
+ * @param table The table device view used for evaluation.
+ * @param literals Array of literal values used for evaluation.
+ * @param output_column The output column where results are stored.
+ * @param data_references Array of data references.
+ * @param operators Array of operators to perform.
+ * @param operator_source_indices Array of source indices for the operators.
+ * @param num_operators Number of operators.
+ * @param num_intermediates Number of intermediates, used to allocate a portion of shared memory to
+ * each thread.
+ */
+template <size_type block_size>
+__launch_bounds__(block_size) __global__
+  void compute_column_kernel(table_device_view const table,
+                             const cudf::detail::fixed_width_scalar_device_view_base* literals,
+                             mutable_column_device_view output_column,
+                             const detail::device_data_reference* data_references,
+                             const ast_operator* operators,
+                             const cudf::size_type* operator_source_indices,
+                             cudf::size_type num_operators,
+                             cudf::size_type num_intermediates);
+
+struct ast_plan {
```

Comment:
This could still use docs? (Why was it resolved?)

Reply:
@harrism My mistake. I had added docs in commit d8479b4b4f5e11dd4573f7af54654f798b613e8b but I had made mistakes in implementation (the host buffer didn’t live long enough for the async copy, I believe) and I reverted that commit. I marked this as resolved before realizing my error.

Reply:
Is this still TODO?

Reply:
Yes, this should be resolved in conjunction with the conversation above. The docstring I wrote previously (and reverted) is copied below, but will probably need to be edited depending on the final behavior of this class. Currently, the `ast_plan` class does not actually perform the host-device copy as described below.

```c++
/**
 * @brief The AST plan creates a device buffer of data needed to execute an AST.
 *
 * On construction, an AST plan creates a single "packed" host buffer of all necessary data arrays,
 * and copies that to the device with a single host-device memory copy. Because the plan tends to be
 * small, this is the most efficient approach for low latency. The stream is not synchronized
 * automatically, so a stream sync must be performed manually (or by another function) before the
 * device data can be used safely.
 *
 */
```